### ISSUE TITLE
Improve build performance

### ISSUE THREAD

**Issue by laeubi** [eclipse-platform/eclipse.platform#276]
Currently a verification build take around 30 minutes. This issue is meant to collect some stats and ideas to improve build performance. Once done, this might be applied for other platform projects as well.

**Comment by laeubi** [comment]
Build was run on a 16 CPU machine with the "smart builder" enabled this gives the following statistics:

```
[INFO] Task segment [clean, verify], number of projects 75
[INFO] Segment walltime 579 s, segment projects service time 2124 s, effective/maximum degree of concurrency 3,67/16
```
That indicates that we could probably get already faster builds with more CPUs, I have opened https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/2277 to investigate if we can get more CPU power.

Enabling builder profiling currently shows the following bottlenecks:
```
[INFO] Segment walltime 455 s, segment projects service time 1360 s, effective/maximum degree of concurrency 2,99/16
[INFO] Build critical path service times (and bottleneck** times):
   org.eclipse.platform:eclipse.platform                            0 s (    0 s)
   org.eclipse.platform:eclipse.platform.runtime                    0 s
   org.eclipse.platform:runtime-bundles                             0 s
   org.eclipse.platform:org.eclipse.core.contenttype               32 s
   org.eclipse.platform:org.eclipse.core.runtime                   27 s
   org.eclipse.platform:org.eclipse.core.expressions               44 s
   org.eclipse.platform:org.eclipse.core.resources                 30 s (    9 s)
   org.eclipse.platform:org.eclipse.team.core                      31 s
   org.eclipse.platform:org.eclipse.team.ui                        17 s (   10 s)
   org.eclipse.platform:org.eclipse.core.tests.resources          248 s (  143 s)
   org.eclipse.platform:org.eclipse.compare.tests                  15 s (   11 s)
   org.eclipse.team:org.eclipse.team.tests.core                     5 s (    5 s)
Build critical path total service time   455 s
Build bottleneck projects service times (and bottleneck** times):
   org.eclipse.platform:org.eclipse.core.tests.resources          248 s (  143 s)
   org.eclipse.platform:org.eclipse.compare.tests                  15 s (   11 s)
   org.eclipse.platform:org.eclipse.team.ui                        17 s (   10 s)
   org.eclipse.platform:org.eclipse.core.resources                 30 s (    9 s)
   org.eclipse.team:org.eclipse.team.tests.core                     5 s (    5 s)
   org.eclipse.platform:eclipse.platform                            0 s (    0 s)
Build bottlenecks total time   180 s
** Bottlenecks are projects that limit build concurrency
   removing bottlenecks improves overall build time
```

**Comment by laeubi** [comment]
Maven profiler output shows that API analysis can take considerable amount of time for example:

<h2>[bundle] Eclipse Contexts (53,16 s)</h2>
            

Plugin execution | Duration
-- | --
org.eclipse.tycho.extras:tycho-eclipserun-plugin:4.0.0-SNAPSHOT:eclipse-run {execution: api-analysis} | 49,90 s
org.eclipse.tycho.extras:tycho-eclipserun-plugin:4.0.0-SNAPSHOT:eclipse-run {execution: build-.api_description} | 2,864 s

In this case 52,764 seconds of a total of 53,16 seconds are spend for API analysis tasks!

**Comment by mickaelistria** [comment]
I believe most of the time spent on the API analysis is about provisioning the Target Platform that is used as baseline. We can probably achieve it better, by downloading the SDK once in a preliminary step and using it as baseline when available; that would save some downloads. Or we can maybe improve the locations being used to ensure that same bundle pools are used and artifacts get downloaded only once.

**Comment by vogella** [comment]
@laeubi found something here to improve the build performance?

**Comment by laeubi** [comment]
@vogella yes I did some investigations but not yet had time to further work on this because of other work, is there any concrete issue at the momment?

**Comment by laeubi** [comment]
@vogella FYI current build time (even with the recent merged repos) is now below 10 minutes in github action and below 15 minutes in Jenkins after recent optimizations, I therefore will close this as done for now, if build times increase in the future (e.g. more repos are merged) we can open a new ticket and investigate further on this.

Some things not yet optimized:

- [ ] API analysis might be skipped if the bundle is replaced by the baseline version
- [ ] investigate if we can get better aching on the downloaded artifacts, e.g. if there is a persistent cache folder like for the GH action
- [ ] More CPU power, currently only two CPUs / build are used
